python/ray/tune/ray_trial_executor.py

```diff
@@ -944,6 +877,150 @@ def _change_working_directory(self, trial):
         else:
             yield
 
+    def get_next_executor_event(
+        self, live_trials: Set[Trial], next_trial_exists: bool
+    ) -> ExecutorEvent:
+        """Get the next executor event to be processed in TrialRunner.
+
+        In case there are multiple events available for handling, the next
+        event is determined by the following priority:
+        1. if there is `next_trial_exists`, and if there is cached resources
+        to use, PG_READY is emitted.
+        2. if there is `next_trial_exists` and there is no cached resources
+        to use, wait on pg future and randomized other futures. If multiple
+        futures are ready, pg future will take priority to be handled first.
+        3. if there is no `next_trial_exists`, wait on just randomized other
+        futures.
+
+        An example of #3 would be synchronous hyperband. Although there are pgs
+        ready, the scheduler is holding back scheduling new trials since the
+        whole band of trials is waiting for the slowest trial to finish. In
+        this case, we prioritize handling training result to avoid deadlock
+        situation.
+
+        This is a blocking wait with a timeout (specified with env var).
+        The reason for the timeout is
+        we still want to print status info periodically in TrialRunner for
+        better user experience.
+
+        The handle of `ExecutorEvent.STOP_RESULT` is purely internal to
+        RayTrialExecutor itself. All the other future results are handled by
+        TrialRunner.
+
+        In the future we may want to do most of the handle of
+        `ExecutorEvent.RESTORE_RESULT` and `SAVING_RESULT` in
+        RayTrialExecutor itself and only notify TrialRunner to invoke
+        corresponding callbacks. This view is more consistent with our goal
+        of TrialRunner responsible for external facing Trial state transition,
+        while RayTrialExecutor responsible for internal facing transitions,
+        namely, `is_saving`, `is_restoring` etc.
+
+        Also you may notice that the boundary between RayTrialExecutor and
+        PlacementGroupManager right now is really blurred. This will be
+        improved once we move to an ActorPool abstraction.
```

Comment:
What will be changed with the introduction of this abstraction?

Reply:
A few improvement:
- All resource related internal data structures should be delegated to ActorPool or any successor of PG Manager, instead of spreading across RayTrialExecutor and PGManager. For example, `RayTrialExecutor._cached_actor_pg`, `PGManager._cached_pgs` alike should be in one place. This is easier to keep consistency (so that we request exactly what we need...)
- Bullet proof support of heterogeneous trials. Our current code mostly likely is not going to work in its current state.
- Unit testing... we don't have test for placement group manager, partially because we don't have a good API. Test today is done with Executor together.

There are more details here: https://docs.google.com/document/d/1RDsW7SVzwMPZfA0WLOPA4YTqbRyXIHGYmBenJk33HaE/edit#heading=h.rb8vp3b62vu2

Reply:
OK perfect. Makes sense.